File: 746-min-cost-climbing-stairs/746-min-cost-climbing-stairs.java
class Solution {
    private HashMap<Integer, Integer> memo = new HashMap<Integer, Integer>();

    public int minCostClimbingStairs(int[] cost) {
        return minimumCost(cost.length, cost);
    }

    private int minimumCost(int i, int[] cost) {
        // Base case, we are allowed to start at either step 0 or step 1
        if (i <= 1) {
            return 0;
        }

        // Check if we have already calculated minimumCost(i)
        if (memo.containsKey(i)) {
            return memo.get(i);
        }

        // If not, cache the result in our hash map and return it
        int downOne = cost[i - 1] + minimumCost(i - 1, cost);
        int downTwo = cost[i - 2] + minimumCost(i - 2, cost);
        memo.put(i, Math.min(downOne, downTwo));
        return memo.get(i);
    }
}

File: 746-min-cost-climbing-stairs/746-min-cost-climbing-stairs.cpp
class Solution {
public:
    int minCostClimbingStairs(vector<int>& cost) {
        vector<int> dp(cost.size(),0);
        dp[0] = cost[0];
        dp[1] = cost[1];
        int result = 0;
        for(int i = 2; i < cost.size(); i++){
            dp[i] = min(dp[i-2],dp[i-1]) + cost[i];
        }
        return min(dp[cost.size()-1],dp[cost.size()-2]);
    }
};